src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -202,33 +213,92 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
+            var tcs = new TaskCompletionSource<Socket>();
+            var lastError = new StrongBox<Exception>();
+            var pendingCompleteCount = new StrongBox<int>(serverAddresses.Length);
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                // Start all connection tasks now, to prevent possible race conditions with
+                // calling ConnectAsync on disposed sockets.
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    var socket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                    Task connectTask = socket.ConnectAsync(address, port);
+
+                    sockets.Add(socket);
+                    connectTasks.Add(connectTask);
                 }
-                catch (Exception exc)
+                catch (Exception e)
                 {
-                    socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    Interlocked.CompareExchange(ref lastError.Value, e, null);
                 }
             }
 
-            // Propagate the last failure that occurred
-            if (lastException != null)
+            Debug.Assert(sockets.Count == connectTasks.Count, "Expected sockets and tasks lists to have equal size");
+            for (int i = 0; i < sockets.Count; i++)
+            {
+                ParallelConnectHelper(sockets[i], connectTasks[i], tcs, pendingCompleteCount, lastError, sockets);
+            }
+
+            return tcs.Task;
+        }
+
+        private static async void ParallelConnectHelper(
+            Socket socket,
+            Task connectTask,
+            TaskCompletionSource<Socket> tcs,
+            StrongBox<int> pendingCompleteCount,
+            StrongBox<Exception> error,
+            List<Socket> sockets)
+        {
+            try
+            {
+                bool success = false;
+
+                // Try to connect.  If we're successful, store this task into the result task.
+                await connectTask.ConfigureAwait(false);
+                if (connectTask.IsFaulted)
+                {
+                    Interlocked.CompareExchange(ref error.Value, connectTask.Exception, null);
+                }
+                else if (socket.Connected)
+                {
+                    success = tcs.TrySetResult(socket);
+                }
+
+                if (success)
+                {
+                    // Cleanup in finally triggers exactly at 0, so disable with -1.
+                    Interlocked.Exchange(ref pendingCompleteCount.Value, -1);
+
+                    sockets.Remove(socket);
+                    foreach (Socket leftoverSocket in sockets)
```

Comment:
The ObjectDisposed exception was being thrown from calling ConnectAsync on the socket, not during task execution.